### ISSUE TITLE
KeyError on lambda in function

### ISSUE THREAD

**Issue by Murlocks** [bpython/bpython#573]
Reproducible by:
```def foo(s):
    l = map(lambd*CRASH*```

Traceback (most recent call last):
  File "/usr/bin/bpython-curses", line 9, in <module>
    load_entry_point('bpython==0.14.1', 'console_scripts', 'bpython-curses')()
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 1961, in main
    banner=banner)
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 1843, in curses_wrapper
    return func(stdscr, *args, **kwargs)
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 1928, in main_curses
    exit_value = clirepl.repl()
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 1126, in repl
    inp = self.get_line()
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 670, in get_line
    if self.p_key(key) is None:
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 1021, in p_key
    self.addstr(key)
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 369, in addstr
    self.complete()
  File "/usr/lib/python3.5/site-packages/bpython/cli.py", line 461, in complete
    list_win_visible = repl.Repl.complete(self, tab)
  File "/usr/lib/python3.5/site-packages/bpython/repl.py", line 599, in complete
    history=self.history)
  File "/usr/lib/python3.5/site-packages/bpython/autocomplete.py", line 466, in get_completer_bpython
    return get_completer(BPYTHON_COMPLETER, cursor_offset, line, **kwargs)
  File "/usr/lib/python3.5/site-packages/bpython/autocomplete.py", line 446, in get_completer
    cursor_offset, line, **kwargs)
  File "/usr/lib/python3.5/site-packages/bpython/autocomplete.py", line 421, in matches
    history=history)
  File "/usr/lib/python3.5/site-packages/bpython/autocomplete.py", line 376, in matches
    completions = script.completions()
  File "/usr/lib/python3.5/site-packages/jedi/api/__init__.py", line 185, in completions
    user_stmt = self._parser.user_stmt_with_whitespace()
  File "/usr/lib/python3.5/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/lib/python3.5/site-packages/jedi/parser/user_context.py", line 303, in user_stmt_with_whitespace
    user_stmt = self.user_stmt()
  File "/usr/lib/python3.5/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/lib/python3.5/site-packages/jedi/parser/user_context.py", line 293, in user_stmt
    module = self.module()
  File "/usr/lib/python3.5/site-packages/jedi/parser/user_context.py", line 339, in module
    return self._parser().module
  File "/usr/lib/python3.5/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/lib/python3.5/site-packages/jedi/parser/user_context.py", line 283, in _parser
    parser = FastParser(self._grammar, self._source, self._path)
  File "/usr/lib/python3.5/site-packages/jedi/parser/fast.py", line 109, in __call__
    p.update(source)
  File "/usr/lib/python3.5/site-packages/jedi/parser/fast.py", line 257, in update
    self._parse(source)
  File "/usr/lib/python3.5/site-packages/jedi/parser/fast.py", line 383, in _parse
    line_offset, nodes)
  File "/usr/lib/python3.5/site-packages/jedi/parser/fast.py", line 435, in _get_node
    p = Parser(self._grammar, parser_code, self.module_path, tokenizer=tokenizer)
  File "/usr/lib/python3.5/site-packages/jedi/parser/__init__.py", line 155, in __init__
    self.module = p.parse(self._tokenize(tokenizer))
  File "/usr/lib/python3.5/site-packages/jedi/parser/pgen2/parse.py", line 106, in parse
    if self.addtoken(type, value, prefix, start_pos):
  File "/usr/lib/python3.5/site-packages/jedi/parser/pgen2/parse.py", line 170, in addtoken
    value, start_pos, prefix, self.addtoken)
  File "/usr/lib/python3.5/site-packages/jedi/parser/__init__.py", line 269, in error_recovery
    self._stack_removal(grammar, stack, index + 1, value, start_pos)
  File "/usr/lib/python3.5/site-packages/jedi/parser/__init__.py", line 304, in _stack_removal
    clear_names(children=node[1])
  File "/usr/lib/python3.5/site-packages/jedi/parser/__init__.py", line 298, in clear_names
    self._scope_names_stack[-1][c.value].remove(c)
KeyError: 'l'

**Comment by sebastinas** [comment]
Ah, one of the jedi erros. This should be fixed in master. We should probably cherry pick the corresponding commits and release a 0.14.2.

**Comment by pirate** [comment]
Not sure if it's related, but you can trigger a bpython/Jedi crash by pasting the following in bpython:
```python
funcs = {
    'mul': lambda x, y: lambda: x * y,
}
```
```python
  File "/usr/local/lib/python2.7/site-packages/bpython/curtsiesfrontend/repl.py", line 878, in update_completion
    self.list_win_visible = BpythonRepl.complete(self, tab)
  File "/usr/local/lib/python2.7/site-packages/bpython/repl.py", line 599, in complete
    history=self.history)
  File "/usr/local/lib/python2.7/site-packages/bpython/autocomplete.py", line 466, in get_completer_bpython
    return get_completer(BPYTHON_COMPLETER, cursor_offset, line, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/bpython/autocomplete.py", line 446, in get_completer
    cursor_offset, line, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/bpython/autocomplete.py", line 421, in matches
    history=history)
  File "/usr/local/lib/python2.7/site-packages/bpython/autocomplete.py", line 376, in matches
    completions = script.completions()
  File "/usr/local/lib/python2.7/site-packages/jedi/api/__init__.py", line 185, in completions
    user_stmt = self._parser.user_stmt_with_whitespace()
  File "/usr/local/lib/python2.7/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/user_context.py", line 303, in user_stmt_with_whitespace
    user_stmt = self.user_stmt()
  File "/usr/local/lib/python2.7/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/user_context.py", line 293, in user_stmt
    module = self.module()
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/user_context.py", line 339, in module
    return self._parser().module
  File "/usr/local/lib/python2.7/site-packages/jedi/cache.py", line 143, in wrapper
    result = func(self)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/user_context.py", line 283, in _parser
    parser = FastParser(self._grammar, self._source, self._path)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/fast.py", line 109, in __call__
    p.update(source)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/fast.py", line 257, in update
    self._parse(source)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/fast.py", line 383, in _parse
    line_offset, nodes)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/fast.py", line 435, in _get_node
    p = Parser(self._grammar, parser_code, self.module_path, tokenizer=tokenizer)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/__init__.py", line 155, in __init__
    self.module = p.parse(self._tokenize(tokenizer))
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/pgen2/parse.py", line 106, in parse
    if self.addtoken(type, value, prefix, start_pos):
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/pgen2/parse.py", line 170, in addtoken
    value, start_pos, prefix, self.addtoken)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/__init__.py", line 269, in error_recovery
    self._stack_removal(grammar, stack, index + 1, value, start_pos)
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/__init__.py", line 304, in _stack_removal
    clear_names(children=node[1])
  File "/usr/local/lib/python2.7/site-packages/jedi/parser/__init__.py", line 298, in clear_names
    self._scope_names_stack[-1][c.value].remove(c)
KeyError: u'funcs'
```

**Comment by thomasballinger** [comment]
Either fixed in 0.14.2 or no longer crashes due to Jedi fixing it.
